checkmarx/util.py

```diff
@@ -1,205 +1,400 @@
-    ):
-        return True
-    else:
-        return False
 
+def scan_sources(
+    component_descriptor_path: str,
+    cx_client: checkmarx.client.CheckmarxClient,
+    team_id: str,
+    threshold: int,
+    max_workers: int = 8,
+    exclude_paths: typing.Sequence[str] = [],
+    include_paths: typing.Sequence[str] = [],
+) -> model.FinishedScans:
+    component_descriptor = cm.ComponentDescriptor.from_dict(
+        ci.util.parse_yaml_file(component_descriptor_path)
+    )
 
-def is_scan_necessary(project: checkmarx.model.ProjectDetails, hash: str):
-    remote_hash = project.get_custom_field(checkmarx.model.CustomFieldKeys.HASH)
-    if remote_hash != hash:
-        print(f'{remote_hash=} != {hash=} - scan required')
-        return True
-    else:
-        return False
+    components = tuple(product.v2.components(component_descriptor_v2=component_descriptor))
 
+    # identify scan artifacts and collect them in a sequence
+    artifacts_gen = _get_scan_artifacts_from_components(
+        components=components,
+    )
 
-def get_scan_info_table(
-        scan_results: typing.Iterable[checkmarx.model.ScanResult],
-        tablefmt: str = 'simple',
-):
-    scan_info_header = ('ScanId', 'ComponentName', 'ScanState', 'Start', 'End')
-
-    def started_on(scan_result):
-        return scan_result.scan_response.dateAndTime.startedOn if \
-            scan_result.scan_response else 'unknown'
-
-    def ended_on(scan_result):
-        return scan_result.scan_response.dateAndTime.finishedOn if \
-            scan_result.scan_response else 'unknown'
-
-    scan_info_data = (
-        (
-            scan_result.scan_response.id,
-            scan_result.component.name,
-            scan_result.scan_response.status.name,
-            started_on(scan_result),
-            ended_on(scan_result),
-        ) for scan_result in scan_results
+    return scan_artifacts(
+        cx_client=cx_client,
+        exclude_paths=exclude_paths,
+        include_paths=include_paths,
+        max_workers=max_workers,
+        scan_artifacts=artifacts_gen,
+        team_id=team_id,
+        threshold=threshold,
     )
-    scan_info = tabulate.tabulate(
-        headers=scan_info_header,
-        tabular_data=scan_info_data,
-        tablefmt=tablefmt,
+
+
+def _get_scan_artifacts_from_components(
+    components: typing.List[cm.Component],
+) -> typing.Generator:
+    for component in components:
+        for source in component.sources:
+            if source.type is not cm.SourceType.GIT:
+                raise NotImplementedError
+
+            if source.access.type is not cm.AccessType.GITHUB:
+                raise NotImplementedError
+
+            cx_label = get_source_scan_label_from_labels(source.labels)
+
+            if not cx_label or cx_label.policy is sdo.labels.ScanPolicy.SCAN:
+                yield model.ScanArtifact(
+                    access=source.access,
+                    # name=f'{component.name}_{str(source.identity())}',
```

Comment:
```suggestion
```

Reply:
I think line 77 should be used, rather than line 78, right?

Reply:
correct. I changed this